en/lessons/basics/containers.md

```diff
@@ -0,0 +1,752 @@
+
+## When to use Maps
+
+Maps are really great for in memory persistence of state that will need to be
+retrieved by a key of some arbitrary type. This is is because maps have great
+lookup asymptotics (`𝛰(log n)`) due to the ordering on the key values. The
+example that immediately springs to mind is session storage on the server in a
+web application. The session state can be indexed by an id that is stored in
+the cookie, and the session state can be retrieved from an in memory `Map` on
+every request. This solution doesn't scale infinitely, but you would be
+surprised how well it works!
+
+## HashMaps
+
+There are some cases where we do not have an ordering on our type, but still
+want to use it as a key to index a map. In this case we probably want to reach
+for a hashmap. A hashmap simply hashes the key, et voilà, we have an ordering!
+This does require that the key type is hashable though.
+
+The module that exports the `HashMap` data type and functionality is called
+`Data.HashMap.Strict`, it lives in the `unordered-containers` package. The api
+is identical to `Map` aside from a `Hashable k` constraint instead of an `Ord k`
+on the key.
+
+## Vectors
+
+Sometimes we want a list like container where we can have performant indexed
+access. In fact arrays are often the primitive sequential data structure in
+lots of popular programming langauges. Haskell has this kind of data structures
+too, and the most popular implementation is the `Vector` type from the `vector`
+library.
+
+One of the coolest things about the `Vector` type is that it offers `𝛰(1)`
+indexed access. The `vector` library offers both a safe and unsafe index
+accessor operator.
+
+Here is an example where we want to create a container that allows us to
+retrieve ascii chars by their character code.
+
+__Note__: You can fetch the `vector` library using `cabal repl --build-depends "vector"`
+or `stack exec --package vector -- ghci`
+
+```haskell
+ghci> import qualified Data.Vector as V
+ghci> asciiChars = V.fromList ['\NUL'..'\DEL']
+ghci> asciiChars ! 48
+'0'
+ghci> asciiChars ! 97
+'a'
+ghci> asciiChars ! 65
+'A'
+ghci> asciiChars !? 65
+Just 'A'
+ghci> asciiChars !? 128
+Nothing
+ghci> asciiChars !? 127
+Just '\DEL'
+```
+
+Vectors also offer efficient slicing `𝛰(1)`, so they are really good for
+creating sub-vectors. The slice function takes the index to start at, the
+length of the slice, and the vector to slice, and returns a new vector. Slicing
+is unsafe, if the starting index plus the length of the slice exceeds the vector
+you will get a runtime error.
+
+```haskell
+ghci> :t V.slice
+V.slice :: Int -> Int -> V.Vector a -> V.Vector a
+ghci> lowerCase = V.slice 97 26 asciiChars
+ghci> lowerCase
+"abcdefghijklmnopqrstuvwxyz"
+ghci> upperCase = V.slice 65 26 asciiChars
+ghci> upperCase
+"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+ghci> nums = V.slice 48 10 asciiChars
+ghci> nums
+"0123456789"
+ ghci> V.slice 97 92 asciiChars -- Error case
```

Comment:
I feel like you could say a bit more than "Error case". Like "Here, we attempt <forbidden operation>, and it fails"